lib/assert.js

```diff
@@ -931,6 +933,88 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+class CallTracker {
+
+  constructor() {
+    this.callChecks = new SafeSet();
+  }
+
+  calls(fn, exact = 1) {
+    if (process._exiting)
+      throw new ERR_UNAVAILABLE_DURING_EXIT();
+    if (typeof fn === 'number') {
+      exact = fn;
+      fn = noop;
+    } else if (fn === undefined) {
+      fn = noop;
+    }
+
+    validateUint32(exact, 'exact', true);
+
+    const context = {
+      exact,
+      actual: 0,
+      // eslint-disable-next-line no-restricted-syntax
+      stackTrace: new Error(),
+      name: fn.name || 'calls'
+    };
+    this.callChecks.add(context);
+    const callChecks = this.callChecks;
+
+    return function() {
+      // Increment call count
+      context.actual++;
+      if (context.actual === context.exact) {
+        // Once function has reached its call count remove it from
+        // callChecks set to prevent memory leaks
+        callChecks.delete(context);
+      }
+      // If function has been called more than expected times, add back into
+      // callchecks.
+      if (context.actual === context.exact + 1) {
+        callChecks.add(context);
+      }
+      return fn.apply(this, arguments);
+    };
+  }
+
+  report() {
+    // Create array for errors
+    const errors = [];
+    for (const context of this.callChecks) {
+      // If functions have not been called exact times
+      if (context.actual !== context.exact) {
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+        errors.push({
+          message,
+          actual: context.actual,
+          expected: context.exact,
+          operator: context.name,
+          stack: context.stackTrace
+        });
+      }
+    }
+    return errors;
+  }
+
+  verify() {
+    const errors = this.report();
+    if (errors.length > 0) {
+      const err = new AssertionError({
+        message: 'Function(s) were not called the expected number of times',
+        details: errors,
+      });
+      throw err;
+    }
+  }
+}
+
+assert.CallTracker = CallTracker;
+
```

Comment:
Nit: could we also add a utility function to auto-register the `verify` on the process.exit?
i.e.
```js
verifyOnExit() {
  process.on('exit', () => this.verify());
}
```

Reply:
Yeah I like this idea